tests/failpoints/cases/test_pending_peers.rs

```diff
@@ -28,13 +28,96 @@ fn test_pending_peers() {
 
     fail::cfg(region_worker_fp, "sleep(2000)").unwrap();
     pd_client.must_add_peer(region_id, new_peer(3, 3));
-    sleep(Duration::from_secs(1));
+    sleep_ms(1000);
     let pending_peers = pd_client.get_pending_peers();
     // Region worker is not started, snapshot should not be applied yet.
     assert_eq!(pending_peers[&3], new_peer(3, 3));
     // But it will be applied finally.
     must_get_equal(&cluster.get_engine(3), b"k1", b"v1");
-    sleep(Duration::from_millis(100));
+    sleep_ms(100);
     let pending_peers = pd_client.get_pending_peers();
     assert!(pending_peers.is_empty());
 }
+
+#[test]
+fn test_pending_snapshot() {
+    let _guard = crate::setup();
+    let mut cluster = new_node_cluster(0, 3);
+    configure_for_snapshot(&mut cluster);
+    let election_timeout = configure_for_lease_read(&mut cluster, None, Some(15));
+    let gc_limit = cluster.cfg.raft_store.raft_log_gc_count_limit;
+    cluster.cfg.raft_store.pd_heartbeat_tick_interval = ReadableDuration::millis(100);
+
+    let handle_snapshot_fp = "apply_on_handle_snapshot_1_1";
+    let handle_snapshot_finish_fp = "apply_on_handle_snapshot_finish_1_1";
+    fail::cfg("apply_on_handle_snapshot_sync", "return").unwrap();
+
+    let pd_client = Arc::clone(&cluster.pd_client);
+    // Disable default max peer count check.
+    pd_client.disable_default_operator();
+
+    let region_id = cluster.run_conf_change();
+    pd_client.must_add_peer(region_id, new_peer(2, 2));
+    cluster.must_transfer_leader(region_id, new_peer(1, 1));
+    cluster.must_put(b"k1", b"v1");
+
+    fail::cfg(handle_snapshot_fp, "pause").unwrap();
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    // Give some time for peer 3 to request snapshot.
+    sleep_ms(100);
```

Comment:
How about `must_get_equal(engine_3, b"k1", b"v1")`?

Reply:
It may lead to panic because the leader's apply worker has been paused.